registry/storage/revisionstore.go

```diff
@@ -131,16 +160,25 @@ func (rs *revisionStore) link(revision digest.Digest) error {
 	return rs.blobStore.link(revisionPath, revision)
 }
 
-// delete removes the specified manifest revision from storage.
+// delete creates a tombstone for the given digest revision.  The tombstone
+// is a link file which points to the deleted manifest.
 func (rs *revisionStore) delete(revision digest.Digest) error {
-	revisionPath, err := rs.pm.path(manifestRevisionPathSpec{
-		name:     rs.Name(),
-		revision: revision,
-	})
-
+	// don't allow deleting a blob that doesn't already exist
+	exists, err := rs.exists(revision)
 	if err != nil {
 		return err
 	}
 
-	return rs.driver.Delete(rs.repository.ctx, revisionPath)
+	if !exists {
+		return distribution.ErrUnknownManifestRevision{
+			Name:     rs.Name(),
+			Revision: revision,
+		}
+	}
+
+	if err := rs.tomb.putTombstone(rs.repository.ctx, rs.Name(), revision); err != nil {
+		return err
+	}
+
```

Comment:
@stevvooe Indeed it's not easy to make it consistent with the current `tag` implementation because there is a `tag` field in the image manifest.
If the `tag` field is removed from the image manifest, and make `tag` just a reference to the image manifest (like `ln -s ` or the git tag implementation), it should be a little easier to make it consistent.  This needs 2 new APIs: `POST /tags` to create tags and `DELTE /tags/tag-name` to delete tags.